src/epics/chatEpics.js

```diff
@@ -0,0 +1,227 @@
+                        message: ajaxResult.response,
+                        mID,
+                        secretKey: keyPair.secretKey
+                    }))
+                    .mergeMap(createDecryptMessageObservable)
+                    .mergeMap(authToken =>
+                        chatHandler.initConnection({
+                            authToken,
+                            secretKey: keyPair.secretKey,
+                            mID
+                        })
+                    )
+                    .mergeMap(() =>
+                        Observable.merge(
+                            chatHandler.getMessageSubject()
+                            .map(addMessage),
+
+                            chatHandler.getUserStatusSubject()
+                            .map(setUserStatus),
+
+                            Observable.of(connectionInitiated())
+                        ))
+                
+            )
+        )
+        .catch(error => {
+            console.error(error);
+            return Observable.from([
+                // TODO: Add alert message to all custom errors thrown and show that instead.
+                alertWarning('Something is wrong with the chat thingie. Trying to reconnect...'),
+                disconnected()
+            ])
+        });
+
+const setUsernameEpic = (action$) =>
+    action$.ofType(CHAT_SET_USERNAME)
+    .filter(action => action.username)
+    .mergeMap(action => 
+        Observable.of(action.username)
+            .do(action => {
+                const ttl = USER_STATUS_DELAY_MS / 1000;
+                const username = action.username;
+                chatHandler.sendUserStatus(username, 'viewing', ttl)
+            })
+            .retryWhen(error => error.delay(500))
+    )
+    .map(username => usernameSet(username))
+
+const ownUserStatusEpic = (action$, store) =>
+    action$.ofType(CHAT_INIT_CHAT)
+    .mergeMap(() =>
+        createDetectVisibilityObservable()
+        //.throttleTime(USER_STATUS_DELAY_MS)
+        .do(console.log)
+        .filter(() => store.getState().chat.username)
+        .do(status => {
+            const ttl = USER_STATUS_DELAY_MS / 1000;
+            const { chat: { username } } = store.getState();
+            chatHandler.sendUserStatus(username, status, ttl);
+        })
+        // TODO: Move action to its own action creator in chatActions.js
+        .map(() => ({ type: 'CHAT_STATUS_SENT' }))
+    )
+    .catch(error => {
+        console.error(error)
+            // TODO: Handle error and send alert
+        return Observable.of(alertWarning('What is this black magic!?'));
+    });
+
+const sendMessageEpic = action$ =>
+    action$.ofType(CHAT_SEND_MESSAGE)
+    .mergeMap(action =>
+
+        Observable.of(action)
+        .do(action => {
+            chatHandler.sendMessage(action.message, action.username);
+        })
+        .retryWhen(error => error.delay(1000))
+    )
+    .map(() => ({ type: 'XXXXX' }))
```

Comment:
@Spetastian `CHAT_MESSAGE_SENT`, perhaps?

Reply:
Yes, there are some of these placeholders left. I will fix them when i am doing the final touches.

Reply:
:+1: